File: MATLAB/src/2_NonparametricFRF/time2frf_lpm.m
function [FRF,freq,sG,T] = time2frf_lpm(u,y,fs,varargin)
%TIME2FRF_LPM - Local Polynomial Method FRF estimation (SISO/SIMO).
%   [FRF,freq,sG,T] = time2frf_lpm(u,y,fs)                              % broadband
%   [FRF,freq,sG,T] = time2frf_lpm(u,y,fs,'period',N,'lines',ex)        % periodic
%   [...] = time2frf_lpm(...,'order',R,'halfwidth',n,'band',[fl fh])
%
% The Local Polynomial Method (LPM) estimates the FRF together with the
% transient (leakage) term, so it gives low-bias FRF estimates even when a
% long start-up transient is present (no transient period need be discarded).
%
% PERIODIC mode ('period',N,'lines',ex)  -- the proper LPM for periodic data
%   (Pintelon-Schoukens; Schoukens et al.). The full record of P periods of N
%   samples is transformed with one PN-point DFT, P times finer than a single
%   period. The excited lines then sit at every P-th bin; the bins BETWEEN
%   them carry the transient only. Around each excited bin K the response is
%       Y(K+m) = Y0(k)*[m==0] + sum_{r=0}^R t_r m^r ,   m = -n..n  (n<P)
%   i.e. the transient is a local polynomial fitted on the non-excited
%   neighbours and Y0 = G(w_k) U(K) is the spike at the excited bin. Solving
%   the least squares per line gives Ghat = Y0 / U(K).  (Eqs. (8)-(17).)
%
% BROADBAND mode (default) -- consecutive-bin local polynomial on the whole
%   record DFT, for arbitrary / non-periodic excitation.
%
% u,y     : time-domain input (N x 1) and output (N x nrofo)
% fs      : sampling frequency [Hz]
% 'order'     : transient polynomial order R                    (default 2)
% 'halfwidth' : half window n (neighbours each side, n<P)       (default 2)
% 'band'      : [fl fh] frequency band to return                (default full)
% 'period'    : samples per period N (enables periodic mode)
% 'lines'     : excited bin indices ex (1-based into 1..N/2), periodic mode
% FRF,freq,sG,T : FRF, frequency [Hz], FRF std, transient term (t_0)
%
% Reference: R. Pintelon, K. Barbe, G. Vandersteen, J. Schoukens, "Improved
%   (non-)parametric identification of dynamic systems excited by periodic
%   signals", MSSP 25(7) 2683-2704 (2011); Pintelon & Schoukens 2012, Ch. 7.
% Author : Wataru Ohnishi, The University of Tokyo, 2026
%%%%%

% --- options -------------------------------------------------------------
R = 2; n = 2; band = []; Nper = []; lines = [];
for k = 1:2:numel(varargin)
    switch lower(varargin{k})
        case 'order',                 R = varargin{k+1};
        case {'halfwidth','window'},  n = varargin{k+1};
        case 'band',                  band = varargin{k+1};
        case 'period',                Nper = varargin{k+1};
        case 'lines',                 lines = varargin{k+1}(:);
        otherwise
            error('time2frf_lpm:badopt','Unknown option "%s".',varargin{k});
    end
end

[Ntot,nrofi]  = size(u);
[Ny,nrofo]    = size(y);
if Ny ~= Ntot, error('time2frf_lpm:size','u and y must have equal length.'); end
if nrofi ~= 1
    error('time2frf_lpm:mimo', ...
        'Only single-input (SISO/SIMO) data is supported (MIMO LPM is planned).');
end

periodic = ~isempty(Nper) && ~isempty(lines);

if periodic
    % ===== PERIODIC LPM (PN-point DFT, transient from non-excited bins) ====
    P  = floor(Ntot/Nper);                  % number of periods
    if P < 2, error('time2frf_lpm:periods','periodic LPM needs >= 2 periods.'); end
    PN = P*Nper;
    U  = fft(u(1:PN));                       % PN x 1   (energy only at excited bins)
    Y  = fft(y(1:PN,:));                     % PN x nrofo
    exb  = lines;                            % 1-based excited bins of the 1-period grid
    Kidx = P*(exb-1) + 1;                    % 1-based index of those lines in the PN grid
    freqAll = (exb-1)*fs/Nper;
    nl   = numel(exb);
    if n >= P, n = P-1; end                  % neighbours must stay non-excited
    nprm = R + 2;                            % unknowns: Y0 + t_0..t_R
    nw   = 2*n + 1;
    if nw < nprm
        error('time2frf_lpm:window', ...
            'Window too small: 2*halfwidth+1 (=%d) < order+2 (=%d).',nw,nprm);
    end
    powers = 0:R;
    FRF = zeros(nl,nrofo); T = zeros(nl,nrofo); sG = zeros(nl,nrofo);
    for idx = 1:nl
        K  = Kidx(idx);
        m  = (-n:n)';
        rr = min(max(K+m,1),PN);             % PN bin indices (clamped at edges)
        Kr = [double(m==0), m.^powers];      % nw x (R+2): [spike , transient poly]
        Yw = Y(rr,:);
        th = Kr\Yw;                          % (R+2) x nrofo
        FRF(idx,:) = th(1,:)/U(K);           % Y0/U(K)
        T(idx,:)   = th(2,:);                % t_0 (transient at the excited bin)
        q = nw - nprm;
        if q >= 1
            res = Yw - Kr*th;
            c11 = real(diag(inv(Kr'*Kr)));
            s2  = real(sum(conj(res).*res,1))/q;
            sG(idx,:) = sqrt(s2*c11(1))/abs(U(K));
        else
            sG(idx,:) = NaN;
        end
    end
else
    % ===== BROADBAND LPM (consecutive bins of the whole-record DFT) ========
    Mmax = floor(Ntot/2);
    U = fft(u); Y = fft(y);
    Uvec = U(2:Mmax+1);                      % skip DC; bin m -> freq m*fs/Ntot
    Yvec = Y(2:Mmax+1,:);
    freqAll = (1:Mmax)'*fs/Ntot;
    nprm = 2*(R+1); nw = 2*n+1;
    if nw <= nprm
        error('time2frf_lpm:window', ...
            'Window too small: need 2*halfwidth+1 > 2*(order+1).');
    end
    [FRF,T,sG] = lpm_core(Uvec, Yvec, R, n);
end

% --- band selection ------------------------------------------------------
if isempty(band)
    sel = true(size(freqAll));
else
    sel = (freqAll >= band(1)) & (freqAll <= band(2));
end
freq = freqAll(sel);
FRF  = FRF(sel,:);
T    = T(sel,:);
sG   = sG(sel,:);
end

% ===== local helper (broadband consecutive-bin LPM) =========================
function [G,T,sG] = lpm_core(U, Y, R, n)
% Local polynomial of order R, half-window n, over the consecutive grid 1..L,
% modelling Y(k+r) = (sum g_i r^i) U(k+r) + (sum t_i r^i).
L = size(U,1); nrofo = size(Y,2);
np = 2*(R+1);
powers = 0:R;
G = zeros(L,nrofo); T = zeros(L,nrofo); sG = zeros(L,nrofo);
for m = 1:L
    lo = m-n; hi = m+n;
    if lo < 1,  lo = 1;  hi = min(2*n+1,L);     end
    if hi > L,  hi = L;  lo = max(1,L-2*n);      end
    idx = (lo:hi)';
    r   = idx - m;
    Rp  = r.^powers;
    K   = [U(idx).*Rp, Rp];
    Yw  = Y(idx,:);
    th  = K\Yw;
    G(m,:) = th(1,:);
    T(m,:) = th(R+2,:);
    q = numel(idx) - np;
    if q >= 1
        c11 = real(diag(inv(K'*K)));
        res = Yw - K*th;
        s2  = real(sum(conj(res).*res,1))/q;
        sG(m,:) = sqrt(s2*c11(1));
    else
        sG(m,:) = NaN;
    end
end
end

File: fditools/nonparametric/time2frf_lpm.py
"""Local Polynomial Method (LPM) FRF estimation (port of ``time2frf_lpm.m``).

The LPM estimates the FRF together with the transient (leakage) term, giving
low-bias FRF estimates even with a long start-up transient (no transient period
need be discarded).  SISO / SIMO (single input).

* **periodic** mode (``period=N, lines=ex``): a single ``P*N``-point DFT of the
  whole ``P``-period record; the bins between excited lines carry the transient,
  fitted by a local polynomial of order ``R`` over ``2*halfwidth+1`` neighbours.
* **broadband** mode (default): consecutive-bin local polynomial over the whole
  record DFT, for arbitrary / non-periodic excitation.

Reference: Pintelon, Barbe, Vandersteen, Schoukens, MSSP 25(7) 2683-2704 (2011);
Pintelon & Schoukens 2012, Ch. 7.
"""

from __future__ import annotations

import numpy as np

from ..frfdata import FrfData, UserData


def _lpm_mimo_orthogonal(u, y, fs, R, n, band, nrofs, ex):
    """Orthogonal multi-experiment MIMO LPM (transient-free spike + solve).

    u : (N, nu, ne), y : (N, ny, ne) ; ex : 0-based excited bins.
    Returns an ``(ny, nu, nl)`` :class:`FrfData`.
    """
    nu, ne = u.shape[1], u.shape[2]
    ny = y.shape[1]
    df = fs / nrofs
    freq = ex * df
    nl = ex.size
    powers = np.arange(R + 1)

    Y0 = np.zeros((ny, ne, nl), dtype=complex)
    U0 = np.zeros((nu, ne, nl), dtype=complex)
    vY0 = np.zeros((ny, ne, nl))
    for e in range(ne):
        P = u.shape[0] // nrofs
        if P < 2:
            raise ValueError("orthogonal LPM needs >= 2 periods.")
        nn = n if n < P else P - 1
        nw = 2 * nn + 1
        nprm = R + 2
        if nw < nprm:
            raise ValueError(f"window too small: 2*halfwidth+1 (={nw}) < "
                             f"order+2 (={nprm}).")
        PN = P * nrofs
        Ye = np.fft.fft(y[:PN, :, e], axis=0)
        Ue = np.fft.fft(u[:PN, :, e], axis=0)
        Kidx = P * ex                                  # 0-based bins in PN grid
        m = np.arange(-nn, nn + 1)
        Kr = np.hstack([(m == 0).astype(float)[:, None],
                        (m[:, None].astype(float)) ** powers[None, :]])
        c11 = np.real(np.diag(np.linalg.inv(Kr.T @ Kr)))[0]
        q = max(nw - nprm, 1)
        for j in range(nl):
            rr = np.clip(Kidx[j] + m, 0, PN - 1)
            Yw = Ye[rr, :]
            th, *_ = np.linalg.lstsq(Kr, Yw, rcond=None)
            Y0[:, e, j] = th[0, :]
            U0[:, e, j] = Ue[Kidx[j], :]
            res = Yw - Kr @ th
            vY0[:, e, j] = np.real(np.sum(np.conj(res) * res, axis=0)) / q * c11

    G = np.zeros((ny, nu, nl), dtype=complex)
    sG = np.zeros((ny, nu, nl))
    for j in range(nl):
        Um = U0[:, :, j]
        Ym = Y0[:, :, j]
        W = np.linalg.pinv(Um)
        G[:, :, j] = Ym @ W
        for i in range(nu):
            sG[:, i, j] = np.sqrt(vY0[:, :, j] @ np.abs(W[:, i]) ** 2)

    if band is not None:
        sel = (freq >= band[0]) & (freq <= band[1])
        freq, G, sG = freq[sel], G[:, :, sel], sG[:, :, sel]
    ud = UserData(sG=sG, method="lpm", nrofp=ne)
    return FrfData(G, freq, userdata=ud)


def _lpm_mimo_zippered(u, y, fs, R, n, band, nrofs, ex):
    """Single zippered MIMO experiment: per-input SIMO periodic LPM + assemble.

    Each input owns disjoint (interleaved) excited lines; the active input is
    identified per line, the SIMO LPM is run on its owned lines, and every
    column is interpolated onto the full grid (per-channel resolution = 1/nu).
    Useful e.g. for thermal systems (one experiment, several inputs, long
    transient handled by the LPM).  Returns an ``(ny, nu, nl)`` :class:`FrfData`.
    """
    nu = u.shape[1]
    ny = y.shape[1]
    df = fs / nrofs
    freq = ex * df
    nl = ex.size
    P = u.shape[0] // nrofs

    Uavg = np.zeros((nl, nu), dtype=complex)
    for i in range(nu):
        acc = np.zeros(nrofs, dtype=complex)
        for p in range(P):
            acc += np.fft.fft(u[p * nrofs:(p + 1) * nrofs, i])
        Uavg[:, i] = acc[ex] / P
    owner = np.argmax(np.abs(Uavg), axis=1)            # input owning each line

    G = np.full((ny, nu, nl), np.nan, dtype=complex)
    sGm = np.full((ny, nu, nl), np.nan)
    Tm = np.full((ny, nu, nl), np.nan, dtype=complex)
    for i in range(nu):
        idx = np.where(owner == i)[0]
        if idx.size == 0:
            continue
        FRFi, _, sGi, Ti = time2frf_lpm(u[:, i], y, fs, order=R, halfwidth=n,
                                        period=nrofs, lines=ex[idx])
        for o in range(ny):
            G[o, i, idx] = FRFi[:, o]
            sGm[o, i, idx] = sGi[:, o]
            Tm[o, i, idx] = Ti[:, o]

    for i in range(nu):                                # interpolate onto full grid
        fi = np.where(~np.isnan(G[0, i, :]))[0]
        for o in range(ny):
            G[o, i, :] = (np.interp(freq, freq[fi], np.real(G[o, i, fi]))
                          + 1j * np.interp(freq, freq[fi], np.imag(G[o, i, fi])))
            sGm[o, i, :] = np.interp(freq, freq[fi], sGm[o, i, fi])

    if band is not None:
        sel = (freq >= band[0]) & (freq <= band[1])
        freq, G, sGm, Tm = freq[sel], G[:, :, sel], sGm[:, :, sel], Tm[:, :, sel]
    ud = UserData(sG=sGm, T=Tm, method="lpm")
    return FrfData(G, freq, userdata=ud)


def _lpm_core(U, Y, R, n):
    """Broadband consecutive-bin LPM over grid 0..L-1."""
    L = U.shape[0]
    nrofo = Y.shape[1]
    npar = 2 * (R + 1)
    powers = np.arange(R + 1)
    G = np.zeros((L, nrofo), dtype=complex)
    T = np.zeros((L, nrofo), dtype=complex)
    sG = np.zeros((L, nrofo))
    for m in range(L):
        lo, hi = m - n, m + n
        if lo < 0:
            lo, hi = 0, min(2 * n, L - 1)
        if hi > L - 1:
            hi, lo = L - 1, max(0, L - 1 - 2 * n)
        idx = np.arange(lo, hi + 1)
        r = (idx - m).astype(float)
        Rp = r[:, None] ** powers[None, :]              # (nw, R+1)
        K = np.hstack([U[idx][:, None] * Rp, Rp])       # (nw, 2(R+1))
        Yw = Y[idx, :]
        th, *_ = np.linalg.lstsq(K, Yw, rcond=None)
        G[m, :] = th[0, :]
        T[m, :] = th[R + 1, :]
        q = idx.size - npar
        if q >= 1:
            c11 = np.real(np.diag(np.linalg.inv(K.conj().T @ K)))[0]
            res = Yw - K @ th
            s2 = np.real(np.sum(np.conj(res) * res, axis=0)) / q
            sG[m, :] = np.sqrt(np.abs(s2 * c11))   # abs guards edge ill-conditioning
        else:
            sG[m, :] = np.nan
    return G, T, sG


def time2frf_lpm(u, y, fs, order=2, halfwidth=2, band=None,
                 period=None, lines=None):
    """Return ``(FRF, freq, sG, T)``.

    Parameters
    ----------
    u : (N,) or (N, 1)      single-input time data
    y : (N,) or (N, nrofo)  output time data
    fs : float              sampling frequency [Hz]
    order : int             transient polynomial order ``R`` (default 2)
    halfwidth : int         half window ``n`` (neighbours each side; default 2)
    band : (fl, fh), optional   frequency band to return (default: full)
    period : int, optional      samples per period ``N`` -> enables periodic mode
    lines : array_like, optional excited bin indices (**0-based**, into 0..N/2),
        i.e. ``freq = lines * fs / period`` (periodic mode)

    Returns
    -------
    SISO/SIMO : ``(FRF, freq, sG, T)`` arrays.
    MIMO (``u`` of shape ``(N, nu, ne)`` with ``period``/``lines``):
        an ``(ny, nu, nl)`` :class:`FrfData` from the orthogonal
        multiple-experiment solve.
    """
    u = np.asarray(u, dtype=float)
    y = np.asarray(y, dtype=float)
    if u.ndim == 3:                                   # orthogonal MIMO LPM
        if period is None or lines is None:
            raise ValueError("MIMO LPM needs 'period' and 'lines'.")
        return _lpm_mimo_orthogonal(u, y, fs, int(order), int(halfwidth), band,
                                    int(period), np.asarray(lines, dtype=int).ravel())
    if u.ndim == 1:
        u = u[:, None]
    if y.ndim == 1:
        y = y[:, None]
    if u.shape[0] != y.shape[0]:
        raise ValueError("u and y must have equal length.")
    if u.shape[1] > 1:                                # single zippered MIMO exp.
        if period is None or lines is None:
            raise ValueError("zippered MIMO LPM needs 'period' and 'lines'.")
        return _lpm_mimo_zippered(u, y, fs, int(order), int(halfwidth), band,
                                  int(period), np.asarray(lines, dtype=int).ravel())
    Ntot, nrofi = u.shape
    R, n = int(order), int(halfwidth)
    periodic = period is not None and lines is not None

    if periodic:
        Nper = int(period)
        P = Ntot // Nper
        if P < 2:
            raise ValueError("periodic LPM needs >= 2 periods.")
        PN = P * Nper
        U = np.fft.fft(u[:PN, 0])
        Y = np.fft.fft(y[:PN, :], axis=0)
        exb = np.asarray(lines, dtype=int).ravel()       # 0-based 1-period bins
        Kidx = P * exb                                   # 0-based bins in PN grid
        freq_all = exb * fs / Nper
        nl = exb.size
        if n >= P:
            n = P - 1
        nprm = R + 2
        nw = 2 * n + 1
        if nw < nprm:
            raise ValueError(f"window too small: 2*halfwidth+1 (={nw}) < "
                             f"order+2 (={nprm}).")
        powers = np.arange(R + 1)
        nrofo = Y.shape[1]
        FRF = np.zeros((nl, nrofo), dtype=complex)
        T = np.zeros((nl, nrofo), dtype=complex)
        sG = np.zeros((nl, nrofo))
        m = np.arange(-n, n + 1)
        for idx in range(nl):
            K = Kidx[idx]
            rr = np.clip(K + m, 0, PN - 1)
            Kr = np.hstack([(m == 0).astype(float)[:, None],
                            (m[:, None].astype(float)) ** powers[None, :]])
            Yw = Y[rr, :]
            th, *_ = np.linalg.lstsq(Kr, Yw, rcond=None)
            FRF[idx, :] = th[0, :] / U[K]
            T[idx, :] = th[1, :]
            q = nw - nprm
            if q >= 1:
                c11 = np.real(np.diag(np.linalg.inv(Kr.T @ Kr)))[0]
                res = Yw - Kr @ th
                s2 = np.real(np.sum(np.conj(res) * res, axis=0)) / q
                sG[idx, :] = np.sqrt(s2 * c11) / np.abs(U[K])
            else:
                sG[idx, :] = np.nan
    else:
        Mmax = Ntot // 2
        U = np.fft.fft(u[:, 0])
        Y = np.fft.fft(y, axis=0)
        Uvec = U[1:Mmax + 1]
        Yvec = Y[1:Mmax + 1, :]
        freq_all = np.arange(1, Mmax + 1) * fs / Ntot
        nprm = 2 * (R + 1)
        if 2 * n + 1 <= nprm:
            raise ValueError("window too small: need 2*halfwidth+1 > 2*(order+1).")
        FRF, T, sG = _lpm_core(Uvec, Yvec, R, n)

    if band is None:
        sel = np.ones(freq_all.shape, dtype=bool)
    else:
        sel = (freq_all >= band[0]) & (freq_all <= band[1])
    return FRF[sel, :], freq_all[sel], sG[sel, :], T[sel, :]
